dex/src/main/java/io/crate/data/BatchIterator.java

```diff
@@ -0,0 +1,135 @@
+
+
+/**
+ * A iterator used to navigate over batched data.
+ * A iterator usually contains some data already in-memory.
+ * This data can be accessed via {@link #moveNext()} and {@link #currentRow()}.
+ *
+ * Once all in-memory data has been consumed more data can loaded with {@link #loadNextBatch()}
+ * unless {@link #allLoaded()} is true in which case the iterator is exhausted.
+ *
+ * A BatchIterator starts either *before* the first row or in an unloaded state.
+ * This means a consumer can consume a BatchIterator like this:
+ *
+ * <pre>
+ *     while (it.moveNext()) {
+ *         // do something with the row
+ *     }
+ *     if (it.allLoaded()) {
+ *          // iterator is exhausted
+ *     } else {
+ *         it.loadNextBatch().whenComplete((r, t) -> {
+ *             // continue consumption
+ *         }
+ *     }
+ * </pre>
+ *
+ *
+ * Thread-safety notes:
+ *
+ * Concurrent usage of a BatchIterator is not supported.
+ */
+public interface BatchIterator {
+
+    Row OFF_ROW = new Row() {
+        @Override
+        public int numColumns() {
+            throw new IllegalStateException("Iterator is not on a row");
+        }
+
+        @Override
+        public Object get(int index) {
+            throw new IllegalStateException("Iterator is not on a row");
+        }
+
+        @Override
+        public Object[] materialize() {
+            throw new IllegalStateException("Iterator is not on a row");
+        }
+    };
+
+    /**
+     * Moves the Iterator back to the starting position.
+     *
+     * A consumer can then iterate over the rows again by calling into
+     * {@link #moveNext()} and {@link #loadNextBatch()} as appropriate.
+     *
+     * @throws IllegalStateException if the cursor is closed
+     */
+    void moveToStart();
+
+    /**
+     * Advances the iterator.
+     *
+     * @return true if the iterator moved to the next valid row.
+     *         false if the iterator is no longer on a valid row.
+     *         If {@link #allLoaded()} returns true the iterator is out of data. Otherwise {@link #loadNextBatch()}
+     *         can be used to load the next batch - after which {@code moveNext()} can be used again.
+     *
+     * @throws IllegalStateException if the cursor is closed
+     */
+    boolean moveNext();
+
+
+    /**
+     * @return the current row the iterator is positioned on.
+     *         Multiple calls to this method might return the same instance, even after {@link #moveNext()}.
+     *         But underlying data will change.
+     */
+    Row currentRow();
+
+    /**
+     * Closes the iterator and frees all resources.
+     * After this method has been called all methods on the iterator will result in an error.
+     *
+     * There are two exceptions to that:
+     *   - close: Close itself can be called multiple times, but only the first call has an effect.
+     *   - loadNextBatch: This method never raises, but instead returns a failed CompletionStage.
```

Comment:
Because it's not intuativie to have a method raise AND return a (potentially failed) future.
This just puts burden on the consumer because it has to handle the exception in two different ways.

For implementations of the BatchIterator it doesn't really matter.